### PR TITLE
mat: add error handling when asking to slice to 0

### PR DESCRIPTION
**Pull request by ahlusar1989**
Addresses #551 .

WIP - needs tests and clarifications on any other edge cases I have missed from @kortschak. So far I have only one condition to check if i is relative to k and j is relative to l. 

All tests pass and code is formatted.

### CODE REVIEW

File: mat/dense.go [hunk 1]
```diff
@@ -281,6 +281,10 @@ func (m *Dense) rawRowView(i int) []float64 {
 // Slice panics with ErrIndexOutOfRange if the slice is outside the capacity
 // of the receiver.
 func (m *Dense) Slice(i, k, j, l int) Matrix {
+	if i == k || j == l {
+		panic(ErrZeroLength)
+	}
+
 	mr, mc := m.Caps()
 	if i < 0 || mr <= i || j < 0 || mc <= j || k <= i || mr < k || l <= j || mc < l {
```
Comment (kortschak) [inline mat/dense.go]: The condition relating i/k and j/l can be changed to `k < i` and `l < j` since the equality case is handled above.

I'm wondering now about the order of checks though. @btracey What do you think about this? Is an out of range check a higher priority than a zero length? With the change in the para above, that would work.

File: CONTRIBUTORS [hunk 2]
```diff
@@ -71,3 +71,4 @@ Takeshi Yoneda <cz.rk.t0415y.g@gmail.com>
 Tobin Harding <me@tobin.cc>
 Vladimír Chalupecký <vladimir.chalupecky@gmail.com>
 Yevgeniy Vahlis <evahlis@gmail.com>
+Saran Ahluwalia <ahlusar.ahluwalia@gmail.com>
```
Comment (kortschak) [inline CONTRIBUTORS]: Please add yourself to AUTHORS as well.

**Comment by btracey** [comment]
I think it probably doesn't matter which we do first. What we could do, however, is keep the check as is, and then inside the if statement test for equality to throw the right panic. This way there is only one test for correct slice actions.

**Comment by ahlusar1989** [comment]
I'll do just that @btracey . I think @kortschak has an example in the issue #551 . I appreciate all of your feedback.

File: AUTHORS [hunk 3]
```diff
@@ -68,3 +68,5 @@ The University of Washington
 Tobin Harding <me@tobin.cc>
 Vladimír Chalupecký <vladimir.chalupecky@gmail.com>
 Yevgeniy Vahlis <evahlis@gmail.com>
+Saran Ahluwalia <ahlusar.ahluwalia@gmail.com>
```
Comment (btracey) [inline AUTHORS]: Sorry @ahlusar1989 , please add yourself alphabetically in the list (after Sam Zaydel)

File: CONTRIBUTORS (hunk 2 above)
Comment (btracey) [reply CONTRIBUTORS]: Alphabetically, as above.

**Comment by kortschak** [comment]
Can you please add back the final newline.

File: CONTRIBUTORS [hunk 4]
```diff
@@ -70,4 +71,4 @@ Steve McCoy <mccoyst@gmail.com>
 Takeshi Yoneda <cz.rk.t0415y.g@gmail.com>
 Tobin Harding <me@tobin.cc>
 Vladimír Chalupecký <vladimir.chalupecky@gmail.com>
-Yevgeniy Vahlis <evahlis@gmail.com>
+Yevgeniy Vahlis <evahlis@gmail.com>
```
Comment (kortschak) [inline CONTRIBUTORS]: Here too please.
